File: WWF/Tutorials/Tutorial3/Exercise2/Task3/CS/CustomActivity/WebTearActivity.cs
//---------------------------------------------------------------------
//  This file is part of the Windows Workflow Foundation SDK Code Samples.
// 
//  Copyright (C) Microsoft Corporation.  All rights reserved.
// 
//This source code is intended only as a supplement to Microsoft
//Development Tools and/or on-line documentation.  See these other
//materials for detailed information regarding Microsoft code samples.
// 
//THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
//KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
//IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
//PARTICULAR PURPOSE.
//---------------------------------------------------------------------
using System;
using System.ComponentModel;
using System.Workflow.ComponentModel;
using System.Workflow.ComponentModel.Design;

namespace Microsoft.Samples.Workflow.Tutorials.CustomActivity
{
    public class WebTearActivity : System.Workflow.ComponentModel.Activity
    {
        public delegate void PageFinishedEventHandler(object sender, 
            PageFinishedEventArgs e);
        private event PageFinishedEventHandler pageFinishedEvent;
        public event PageFinishedEventHandler PageFinished
        {
            add
            {
                pageFinishedEvent += value;
            }
            remove
            {
                pageFinishedEvent -= value;
            }
        }

        protected override ActivityExecutionStatus Execute
            (ActivityExecutionContext context)
        {
            string pageData;
            System.Net.WebClient client = new System.Net.WebClient();

            try
            {
                // Download the web page data
                pageData = client.DownloadString(this.Url);
            }
            catch (Exception e)
            {
                pageData = e.Message;
            }

            // Raise the PageFinished event back to the host
            pageFinishedEvent(null, new PageFinishedEventArgs(pageData));

            // Notifiy the runtime that the activity has finished
            return ActivityExecutionStatus.Closed;
        }

        public static DependencyProperty UrlProperty =
                    DependencyProperty.Register("Url", 
                    typeof(System.String), 
                    typeof(WebTearActivity));

        [DesignerSerializationVisibilityAttribute
            (DesignerSerializationVisibility.Visible)]
        [BrowsableAttribute(true)]
        [DescriptionAttribute("Url to download")]
        [CategoryAttribute("WebTearActivity Property")]
        public string Url
        {
            get
            {
                return ((string)(base.GetValue(WebTearActivity.UrlProperty)));
            }
            set
            {
                base.SetValue(WebTearActivity.UrlProperty, value);
            }
        }
    }
    
    public class PageFinishedEventArgs
    {
        private string pageData;

        public PageFinishedEventArgs(string data)
        {
            this.pageData = data;
        }

        public string Data
        {
            get { return this.pageData; }
        }
    }
}

File: WWF/Tutorials/Tutorial3/Exercise2/Task1/VB/CustomActivity/WebTearActivity.vb
'---------------------------------------------------------------------
'  This file is part of the Windows Workflow Foundation SDK Code Samples.
' 
'  Copyright (C) Microsoft Corporation.  All rights reserved.
' 
'This source code is intended only as a supplement to Microsoft
'Development Tools and/or on-line documentation.  See these other
'materials for detailed information regarding Microsoft code samples.
' 
'THIS CODE AND INFORMATION ARE PROVIDED AS IS WITHOUT WARRANTY OF ANY
'KIND, EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE
'IMPLIED WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A
'PARTICULAR PURPOSE.
'---------------------------------------------------------------------
Imports Microsoft.VisualBasic
Imports System
Imports System.ComponentModel
Imports System.Workflow.ComponentModel
Imports System.Workflow.ComponentModel.Design

Namespace Microsoft.Samples.Workflow.Tutorials.CustomActivity
    Public Class WebTearActivity : Inherits System.Workflow.ComponentModel.Activity

        Public Delegate Sub PageFinishedEventHandler(ByVal sender As Object, _
            ByVal e As PageFinishedEventArgs)
        Private Event pageFinishedEvent As PageFinishedEventHandler
        Public Custom Event PageFinished As PageFinishedEventHandler
            AddHandler(ByVal value As PageFinishedEventHandler)
                AddHandler pageFinishedEvent, value
            End AddHandler
            RemoveHandler(ByVal value As PageFinishedEventHandler)
                RemoveHandler pageFinishedEvent, value
            End RemoveHandler
            RaiseEvent(ByVal sender As Object, ByVal e As PageFinishedEventArgs)
                RaiseEvent pageFinishedEvent(sender, e)
            End RaiseEvent
        End Event

        Protected Overrides Function Execute(ByVal context As ActivityExecutionContext) _
            As ActivityExecutionStatus
            Dim pageData As String
            Dim client As System.Net.WebClient = New System.Net.WebClient()

            Try
                ' Download the web page data
                pageData = client.DownloadString(Me.Url)
            Catch e As Exception
                pageData = e.Message
            End Try

            ' Raise the PageFinished event back to the host
            RaiseEvent PageFinished(Nothing, New PageFinishedEventArgs(pageData))

            ' Notifiy the runtime that the activity has finished
            Return ActivityExecutionStatus.Closed
        End Function

        Public Shared UrlProperty As DependencyProperty = _
            DependencyProperty.Register("Url", GetType(System.String), _
            GetType(WebTearActivity))

        <DesignerSerializationVisibilityAttribute _
        (DesignerSerializationVisibility.Visible), _
        BrowsableAttribute(True), _
        DescriptionAttribute("Url to download"), _
        CategoryAttribute("WebTearActivity Property")> _
        Public Property Url() As String
            Get
                Return (CStr(MyBase.GetValue(WebTearActivity.UrlProperty)))
            End Get
            Set(ByVal value As String)
                MyBase.SetValue(WebTearActivity.UrlProperty, value)
            End Set
        End Property
    End Class
    
    Public Class PageFinishedEventArgs
        Private pageData As String

        Public Sub New(ByVal data As String)
            Me.pageData = data
        End Sub

        Public ReadOnly Property Data() As String
            Get
                Return Me.pageData
            End Get
        End Property
    End Class
End Namespace
